web/server/go.md

```diff
@@ -0,0 +1,498 @@
+"how do I accomplish the client need with the given dependencies?"
+
+Let's take a look at an example!
+In our dog application,
+it's unlikely all we want to do
+is create dogs in a database.
+Let's say the purpose of the application
+is to show the user how cute a dog is
+given various attributes.
+
+Our handler may look something like this:
+
+```go
+package handlers
+
+type CalculateCutenessScore = func(dogID uuid.UUID) (int, error)
+
+type getCutenessScoreHandler struct {
+    calculateCutenessScore CalculateCutenessScore
+    //logger implementation may vary, we'll psuedocode for this example
+    logger Logger
+}
+
+func (h *getCutenessScoreHandler) handleGetCutenessScore(w http.ResponseWriter, r *http.Request) {
+    // like above we've deferred marshalling for this example
+    p := dogGetParams(w)
+
+    // offload calculation to service function
+    score, err := h.calculateCutenessScore(p.DogID)
+
+    if err == errors.CatchableError {
+        h.logger.Error(fmt.Sprintf("Could not calculate cuteness for dog %v", p.DogID), err)
+        w.Write(cutnessPayload(DEFAULT_CUTENESS_SCORE))
+    } else if err != nil {
+        http.Error(w, "Could not calculate dog's cuteness!", http.StatusInternalServerError)
+        return
+    }
+
+    w.Write(cutenessPayload(score))
+}
+```
+
+The handler here is not too much different than our `GetDog` example above.
+We've changed from `GetDog` to `GetCutenessScore`
+and likewise updated the service definition.
+We've also updated the error catching to be a little more robust.
+In this case,
+we've defined a "catchable error" to demonstrate
+when we should return a default case to the client
+and log an error.
+Then we default to 500 when unknown errors appear.
+
+The service definition is passed into the handler as a dependency,
+but unlike in the original handler definition,
+we've also defined the service contract.
+This is important to understanding services,
+so let's dive in a little deeper.
+
+First let's look at a possible service definition for the cuteness score.
+
+```go
+package services
+
+// NewCalculateCutenessScore return an implementation for calculating dog cuteness
+func NewCalculateCutenessScore(db Database) func(dogID uuid.UUID) (int, error) {
+    return func(dogID uuid.UUID) (int, error) {
+        dog := models.Dog{}
+        err := db.Fetch(&dog, dogID)
+        if err != nil {
+            return 0, err
+        }
+        nameAsInt := strconv.Atoi(dog.Name)
+        // they're all good dogs brent
+        cutenessScore := nameAsInt%10 + BRONT_ADDER
+        return cutenessScore, nil
+    }
+}
+```
+
+So, we've added a new implementation for cuteness score
+to the `services` package.
+It returns a function that can calculate cuteness for a given dog ID.
+The dependencies, like the `Database` connector,
+are passed into the function generator and hidden from the return value.
+
+Notice how this definition is independent of the definition in the handler
```

Comment:
This is the meat of the definition. Please suggest clarifications if necessary.

Reply:
I have some Java-influenced(?) modeling patterns that I usually apply here, but I'm super open to challenging them.

I like to just define an interface for the service in question. Then to have an implementation-specific sub-package that actually implements it (or two...). But this gives consuming code a noun to hold, that has verbs on it. Under the covers Go lets the implementations be closures or whatever (or not) as is useful for your implementation. Also, in the service interface definition layer, I love automatically creating the parallels to `HandlerFunc`, which I believe facilitates unit testing and the closure implementations.

I ack that this sometimes feels ceremonious (a la Java) and wordy, but the flexibility have been really useful for me in the past.

Reply:
MilMove uses a similar layout here: https://github.com/transcom/mymove/tree/master/pkg/services. I don't love the templating it seems to create. I find it's both more go-like and easier to read when the functions are defined directly where they're used. It seems to avoid one or two layers of indirection. I don't know if defining them separate or using a noun has enough benefit to justify those layers. Unless it's meant to be a catalog of sorts.

I think the package structure, like whether to use one or many, is unclear as to best patterns for me. I'd like to experiment with something different.